### PR TITLE
implement kubectl-apply module

### PR DESCRIPTION
**Pull request by sharabiani**
### Kubectl-apply module implemented:
It supports:
- String manifest body
- Local files
- Local template files
- URLs
- Directories containing yaml and template files

### gke-cluster updated and pre-existing-gke-cluster updated:
- To replace kubectl provider with kubectl-apply module to apply create newtrok objects for multinetworking
- To template files as manifests instead of raw content

The template files in gke-cluster and pre-existing-gke-cluster are added to duplicate diff

### Submission Checklist

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cluster Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: modules/management/kubectl-apply/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+## Description
+
+This module simplifies applying Kubernetes manifests to GKE clusters. It provides flexible options for specifying manifests, allowing you to either directly embed them as strings content or reference them from URLs, files, templates, or entire files in directories.
+
+* **Manifest:**
+  * **Raw String:** Specify manifests directly within the module configuration using the `content: manifest_body` format.
+  * **File/Template/Directory Reference:** Set `source` to the path to:
+    * A single URL to a manifest file. Ex.: `https://github.com/.../myrepo/manfiest.yaml`.
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/README.md:8]: nit: typo in name, please rename to 

`https://github.com/.../myrepo/manifest.yaml`.

File: modules/management/kubectl-apply/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    * A single local YAML manifest file (`.yaml`). Ex.: `./manfiest.yaml`.
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/README.md:9]: nit: Typo , please update to 

`./manifest.yaml`.

File: modules/management/kubectl-apply/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  settings:
+    project_id: $(vars.project_id)
+    cluster_name: my-gke-cluster
+    region: us-central1
+
+- id: kubectl-apply
+  source: modules/management/kubectl-apply
+  use: [existing-gke-cluster]
+  settings:
+    - content: "A YAML body"
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/README.md:27]: Can we give real examples here ? 

For example, use 

* `Content` to create a namespace ? 
* `source` to a valid URL , like network object installation
* `template_vars` to have a real map and values ?

File: modules/management/kubectl-apply/kubectl/variables.tf [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+variable "server_side_apply" {
+  description = "Allow using kubectl server-side apply method."
+  type        = bool
+  default     = false
+}
+
+variable "wait_for_rollout" {
+  description = "Wait or not for Deployments and APIService to complete rollout."
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/kubectl/variables.tf:42]: nit: should we add why one would have `wait_for_rollout` as `true` or `false` ? Also if we are always using the value as `true` for this variable, then maybe we should default to `true`

File: modules/management/kubectl-apply/kubectl/variables.tf [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+variable "source_path" {
+  description = "The source for manifest(s) to apply to gke cluster. Acceptable sources are a local yaml or template (.tftpl) file path, a directory (ends with '/') containing yaml or template files, and a url for a yaml file."
+  type        = string
+  default     = null
+}
+
+variable "template_vars" {
+  description = "The values to populate template file(s) with."
+  type        = any
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/kubectl/variables.tf:31]: Can this be of type `map(any)` ?  from https://developer.hashicorp.com/terraform/language/v1.5.x/expressions/type-constraints#map

File: modules/management/kubectl-apply/kubectl/main.tf [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  docs_list = concat(try(local.yaml_body_docs, []), try(data.kubectl_path_documents.yamls[0].documents, []), try(data.kubectl_path_documents.templates[0].documents, []))
+  docs_map = tomap({
+    for index, doc in local.docs_list : index => doc
+  })
+}
+
+data "http" "yaml_content" {
+  count = local.url != null ? 1 : 0
+  url   = local.url
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/kubectl/main.tf:46]: If it is an non valid URL, does terraform throw a good error message ?

File: modules/management/kubectl-apply/kubectl/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  *
+  * Unless required by applicable law or agreed to in writing, software
+  * distributed under the License is distributed on an "AS IS" BASIS,
+  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+output "documents_count" {
+  description = "The count of applying documents."
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/kubectl/outputs.tf:18]: nit: should we name this to `applied_documents_count` ? 

Also a quick question, say the caller intended to apply 10 manifests but the module was only able to apply 9 manifests, at the end of retry operations does it the module automatically revert the successfully applied 9 manifest ?

File: modules/management/kubectl-apply/variables.tf [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  nullable    = false
+}
+
+variable "apply_manifests" {
+  description = "A list of manifests to apply to GKE cluster using kubectl."
+  type = list(object({
+    content           = optional(string, null)
+    source            = optional(string, null)
+    template_vars     = optional(any, null)
+    server_side_apply = optional(bool, false)
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/variables.tf:34]: Would there be ever a condition where `server_side_apply` and `wait_for_rollout` will be separate for each manifest ? In other words, can we extract it to a top level variable of it's own ?

File: modules/management/kubectl-apply/outputs.tf
Comment (ankitkinra) [inline modules/management/kubectl-apply/outputs.tf:1]: This is an empty file, should we delete it ?

File: modules/management/kubectl-apply/providers.tf [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+  * See the License for the specific language governing permissions and
+  * limitations under the License.
+  */
+
+provider "kubectl" {
+  host                   = "https://${data.google_container_cluster.gke_cluster.endpoint}"
+  token                  = data.google_client_config.default.access_token
+  cluster_ca_certificate = base64decode(data.google_container_cluster.gke_cluster.master_auth[0].cluster_ca_certificate)
+  load_config_file       = false
+  apply_retry_count      = 15
```
Comment (ankitkinra) [inline modules/management/kubectl-apply/providers.tf:22]: nit: can you please add a comment here on why we have it set at `15` ?

File: modules/management/kubectl-apply/kubectl/main.tf (hunk 6 above)
Comment (sharabiani) [reply modules/management/kubectl-apply/kubectl/main.tf:46]: For invalid urls there are two condition:

1. If the request cannot be made we'll have:
```
Error: Error making request

  with module.gke_cluster.module.kubectl_apply2.module.kubectl["0"].data.http.yaml_content[0],
  on modules/embedded/modules/management/kubectl-apply/kubectl/main.tf line 44, in data "http" "yaml_content":
  44: data "http" "yaml_content" {

Error making request: GET https: giving up after 1 attempt(s): Get "https:":
http: no Host in request URL
```


2. If the request's response is not a valid yaml doc, then we'll have yaml parsing error:
```
Error: yaml: unmarshal errors:
  line 1: cannot unmarshal !!str `Not Found` into map[string]interface {}

  with module.gke_cluster.module.kubectl_apply2.module.kubectl["0"].kubectl_manifest.apply_doc["0"],
  on modules/embedded/modules/management/kubectl-apply/kubectl/main.tf line 60, in resource "kubectl_manifest" "apply_doc":
  60: resource "kubectl_manifest" "apply_doc" {
```

Comment (ankitkinra) [reply modules/management/kubectl-apply/kubectl/main.tf:46]: That seems like a good error message to me, thanks for testing!

**Review by ankitkinra — APPROVED** [review]

**Pull request merged**